File: composeApp/src/androidMain/kotlin/com/sixclassguys/maplecalendar/ui/calendar/MapleCalendarScreen.kt
package com.sixclassguys.maplecalendar.ui.calendar

import android.os.Build
import androidx.annotation.RequiresApi
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.sixclassguys.maplecalendar.presentation.calendar.CalendarIntent
import com.sixclassguys.maplecalendar.presentation.calendar.CalendarViewModel
import com.sixclassguys.maplecalendar.theme.Typography
import com.sixclassguys.maplecalendar.ui.component.BossScheduleRow
import com.sixclassguys.maplecalendar.ui.component.CalendarCard
import com.sixclassguys.maplecalendar.ui.component.CarouselEventRow
import com.sixclassguys.maplecalendar.ui.component.EmptyEventScreen

@RequiresApi(Build.VERSION_CODES.O)
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MapleCalendarScreen(
    viewModel: CalendarViewModel,
    onNavigateToEventDetail: (Long) -> Unit,
    onNavigateToBossDetail: (Long) -> Unit
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
    val snackbarHostState = remember { SnackbarHostState() }

    LaunchedEffect(uiState.errorMessage) {
        val message = uiState.errorMessage
        if (message != null) {
            snackbarHostState.showSnackbar(message = message)
        }
    }

    Scaffold(
        containerColor = Color.White
    ) { padding ->
        Column(
            modifier = Modifier.fillMaxSize()
                .padding(bottom = padding.calculateBottomPadding())
        ) {

            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(bottom = 32.dp)
            ) {
                // 1. 캘린더 카드
                item {
                    Text(
                        text = "캘린더",
                        style = Typography.titleLarge,
                        modifier = Modifier.padding(16.dp)
                    )

                    CalendarCard(
                        uiState = uiState,
                        onMonthChanged = { newOffset ->
                            viewModel.onIntent(CalendarIntent.ChangeMonth(newOffset))
                        },
                        onDateClick = { date ->
                            viewModel.onIntent(CalendarIntent.SelectDate(date))
                        },
                        today = viewModel.getTodayDate()
                    )
                }

                // 2. 진행중인 이벤트 섹션
                item {
                    val selectedDateText = uiState.selectedDate?.let {
                        "${it.year}년 ${it.monthNumber}월 ${it.dayOfMonth}일"
                    }

                    Text(
                        text = if (selectedDateText == null) "날짜를 선택해주세요!" else "$selectedDateText 진행중인 이벤트",
                        style = Typography.titleMedium,
                        modifier = Modifier.padding(16.dp)
                    )

                    // 해당 월의 이벤트 중 선택된 날짜가 포함된 이벤트 필터링
                    val year = uiState.selectedDate?.year ?: viewModel.getTodayDate().year
                    val month = uiState.selectedDate?.monthNumber ?: viewModel.getTodayDate().monthNumber
                    val day = uiState.selectedDate?.dayOfMonth ?: viewModel.getTodayDate().dayOfMonth
                    val currentKey = "${year}-${month}-${day}"
                    val nowEvents = uiState.eventsMapByDay[currentKey] ?: emptyList()

                    if (nowEvents.isEmpty()) {
                        EmptyEventScreen("진행중인 이벤트가 없어요.")
                    } else {
                        CarouselEventRow(
                            nowEvents = nowEvents,
                            onNavigateToEventDetail = { eventId ->
                                val selected = nowEvents.find { it.id == eventId }
                                selected?.let {
                                    viewModel.onIntent(CalendarIntent.SelectEvent(it.id))
                                    onNavigateToEventDetail(eventId)
                                }
                            }
                        )
                    }
                }

                // 3. 오늘의 보스 일정 (현재 UIState에는 보스 데이터가 없으므로 자리만 유지)
                item {
                    val selectedDateText = uiState.selectedDate?.let {
                        "${it.year}년 ${it.monthNumber}월 ${it.dayOfMonth}일"
                    }

                    Spacer(modifier = Modifier.height(24.dp))
                    Text(
                        text = if (selectedDateText == null) "날짜를 선택해주세요!" else "$selectedDateText 보스 일정",
                        style = Typography.titleMedium,
                        modifier = Modifier.padding(16.dp)
                    )

                    val year = uiState.selectedDate?.year ?: viewModel.getTodayDate().year
                    val month = uiState.selectedDate?.monthNumber ?: viewModel.getTodayDate().monthNumber
                    val day = uiState.selectedDate?.dayOfMonth ?: viewModel.getTodayDate().dayOfMonth
                    val currentKey = "${year}-${month}-${day}"
                    val nowBossSchedules = uiState.bossSchedulesMapByDay[currentKey] ?: emptyList()

                    // 보스 데이터가 비어있을 때 (이미지 포함된 뷰)
                    if (nowBossSchedules.isEmpty()) {
                        EmptyEventScreen("보스 일정이 없어요.")
                    } else {
                        LazyRow(
                            modifier = Modifier.fillMaxSize(),
                            contentPadding = PaddingValues(horizontal = 8.dp)
                        ) {
                            items(nowBossSchedules) { schedule ->
                                BossScheduleRow(
                                    schedule = schedule,
                                    onNavigateToBossDetail = onNavigateToBossDetail
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/presentation/calendar/MapleCalendarScreen.swift
import SwiftUI
import shared

struct MapleCalendarScreen: View {
    
    @StateObject var viewModel: CalendarViewModel
    var onNavigateToEventDetail: (Int64) -> Void
    
    var body: some View {
        
        ZStack {
            Color.white.ignoresSafeArea()
            
            ScrollView(.vertical) {
                VStack(alignment: .leading, spacing: 0) {
                    // 1. 헤더 타이틀
                    Text("캘린더").font(.system(size: 32, weight: .bold))
                        .padding(.horizontal, 16)
                        .padding(.vertical, 16)
                    
                    // 2. 캘린더 카드 섹션
                    CalendarCardView(viewModel: viewModel)
                    
                    // 3. 진행 중인 이벤트 섹션
                    VStack(alignment: .leading, spacing: 16) {
                        let selectedDateText = viewModel.uiState.selectedDate != nil ?
                            "\(viewModel.uiState.selectedDate!.year)년 \(viewModel.uiState.selectedDate!.monthNumber)월 \(viewModel.uiState.selectedDate!.dayOfMonth)일" : "날짜를 선택해주세요!"
                        
                        Text("\(selectedDateText) 진행중인 이벤트")
                            .font(.system(size: 20, weight: .semibold))
                            .padding(.horizontal, 16)
                        
                        let events = viewModel.getSelectedDateEvents()
                        if events.isEmpty {
                            EmptyEventView(message: "진행중인 이벤트가 없어요.")
                        } else {
                            CarouselEventRow(events: events) { id in
                                viewModel.onIntent(intent: .SelectEvent(eventId: id))
                                onNavigateToEventDetail(id)
                            }
                        }
                    }
                    .padding(.top, 24)
                    
                    // 4. 오늘의 보스 일정 섹션
                    VStack(alignment: .leading, spacing: 16) {
                        Text("오늘의 보스 일정")
                            .font(.system(size: 20, weight: .semibold))
                            .padding(.horizontal, 16)
                        
                        // 현재 보스 데이터가 없으므로 Placeholder 처리
                        EmptyEventView(message: "진행중인 이벤트가 없어요.", showBossImage: true)
                    }
                    .padding(.top, 32)
                    .padding(.bottom, 32)
                }
            }
        }
        .onAppear {
            // 상세 페이지에서 돌아왔을 때 선택된 이벤트 상태를 초기화
            viewModel.onIntent(intent: CalendarIntent.SelectEvent(eventId: 0)) // 0 또는 null 처리를 위한 ID
        }
    }
}
